In [36]:
import pandas as pd
import openpyxl
import sys
sys.path.append('../')
from src import sp_eda as se
pd.set_option('display.max_columns', None)

### Insertamos los datos del csv


In [37]:
df1=pd.read_csv('../data/bank-additional.csv')


### Insertamos los datos del xlxs, incorporando las distintas hojas en el mismo dataframe
Cambiamos a su vez el nombre de la columna ID para hacerla coincidir con el del df proveniente del csv

In [39]:
df2012=pd.read_excel('../data/customer-details.xlsx',sheet_name='2012')
df2013=pd.read_excel('../data/customer-details.xlsx',sheet_name='2013')
df2014=pd.read_excel('../data/customer-details.xlsx',sheet_name='2014')

df2=pd.concat([df2012,df2013,df2014],ignore_index=False)
df2_renombrado = df2.rename(columns={'ID': 'id_'})

###  Unimos los dos dataframe 
Eliminamos las columnas Unnamed porvenientes de los dos df originales

In [47]:
df=pd.merge(df1,df2_renombrado,'outer', 'id_')
df=df.drop(columns=['Unnamed: 0_x','Unnamed: 0_y'])

In [48]:
se.eda_prelim(df)

,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth
40559,54.0,management,MARRIED,university.degree,0.0,0.0,0.0,cellular,11.0,21.0,999.0,0.0,NONEXISTENT,1.4,NaN,"-36,1","4,962","5228,1",no,2-junio-2019,33.550,-70.074,f06efe0c-6084-46ae-a820-c175a462f340,22121,1,1,2013-09-14,32
35466,38.0,blue-collar,MARRIED,basic.9y,0.0,1.0,0.0,cellular,161.0,1.0,999.0,0.0,NONEXISTENT,-1.8,"92,893","-46,2","1,291","5099,1",no,18-abril-2019,24.447,-119.656,d246f56d-d2ab-4f7e-8f37-8c913192f73e,53006,2,2,2014-06-08,5
408,32.0,services,MARRIED,high.school,0.0,NaN,NaN,cellular,172.0,1.0,999.0,1.0,FAILURE,-2.9,"92,963","-40,8","1,262","5076,2",no,8-septiembre-2019,49.076,-116.068,02647c16-4a98-413c-b954-d38824f4bda9,10065,0,1,2014-07-10,16
10694,47.0,housemaid,DIVORCED,basic.4y,NaN,0.0,0.0,cellular,357.0,1.0,999.0,0.0,NONEXISTENT,1.4,"93,918","-42,7",NaN,"5228,1",no,28-octubre-2016,33.232,-112.850,3f145a14-64f1-4860-afdc-184dc1b1e89d,135547,0,2,2012-02-27,5
35198,51.0,services,MARRIED,high.school,0.0,0.0,0.0,cellular,1267.0,1.0,999.0,1.0,FAILURE,-1.8,"93,876",-40,"0,668","5008,7",yes,5-junio-2018,30.177,-88.765,d0c8eaed-ce5d-4c93-8e0c-f27673537148,18944,2,2,2014-08-19,4


----------
DIMENSIONES
El conjunto de datos presenta 43170 filas y 28 columnas
----------
INFORMACION DE COLUMNAS
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43170 entries, 0 to 43169
Data columns (total 28 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   age                37880 non-null  float64       
 1   job                42655 non-null  object        
 2   marital            42915 non-null  object        
 3   education          41193 non-null  object        
 4   default            34019 non-null  float64       
 5   housing            41974 non-null  float64       
 6   loan               41974 non-null  float64       
 7   contact            43000 non-null  object        
 8   duration           43000 non-null  float64       
 9   campaign           43000 non-null  float64       
 10  pdays              43000 non-null  float64       
 11  previous           43000 non-null  float64       
 12  po

None

----------
DUPLICADOS
Tenemos 0 duplicados
----------
ELEMENTOS NULOS


age                  12.253880
job                   1.192958
marital               0.590688
education             4.579569
default              21.197591
housing               2.770442
loan                  2.770442
contact               0.393792
duration              0.393792
campaign              0.393792
pdays                 0.393792
previous              0.393792
poutcome              0.393792
emp.var.rate          0.393792
cons.price.idx        1.484827
cons.conf.idx         0.393792
euribor3m            21.834607
nr.employed           0.393792
y                     0.393792
date                  0.968265
latitude              0.393792
longitude             0.393792
id_                   0.000000
Income                0.000000
Kidhome               0.000000
Teenhome              0.000000
Dt_Customer           0.000000
NumWebVisitsMonth     0.000000
dtype: float64

----------
FRECUENCIA CATEGORÍAS
job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
Name: count, dtype: int64
-----------
marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
Name: count, dtype: int64
-----------
education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4y                4356
basic.6y                2386
illiterate                18
Name: count, dtype: int64
-----------
contact
cellular     27396
telephone    15604
Name: count, dtype: int64
-----------
poutcome
NONEXISTENT    37103
FAILURE         4461
SUCCESS         1436
Name: count, dtype: int64
-----------
cons.price.idx
93,994    7938
93,918    6937
92,893    5985
93,444    5349
94,465    4522
93,2      3731
93,075    2552
92,201 

Se observan porcentajes de nulos con el mismo valor en muchas columnas. Se cae en la cuenta de que hemos juntado archivos con distinto numero de registos 43000 vs 43170. los 170 de diferencia sobre 43170 representan el 0,39% mostrado en los elementos nulos. Si esos 170 registros de diferencia no tienen datos de comunicación en la campaña, quizas no tenga sentido incluirlos en el análisis. Se propone una unión de dataframes inner o left 

In [57]:
df=pd.merge(df1,df2_renombrado,'inner', 'id_')
df=df.drop(columns=['Unnamed: 0_x','Unnamed: 0_y'])
se.eda_prelim(df)

,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth
27780,28.0,admin.,SINGLE,high.school,0.0,1.0,0.0,cellular,102,1,999,0,NONEXISTENT,-1.8,"92,843",-50,NaN,"5099,1",no,7-enero-2019,48.702,-109.824,500fdd2b-c32e-4813-9224-23a143e6ada1,24652,1,2,2013-06-10,21
17051,NaN,blue-collar,MARRIED,basic.9y,NaN,1.0,0.0,cellular,560,1,999,0,NONEXISTENT,1.4,"93,918","-42,7",NaN,"5228,1",no,1-octubre-2015,36.002,-105.372,97ceac0b-8081-4ed9-80d5-4de36cb544fe,70469,0,1,2012-03-19,17
36550,31.0,management,MARRIED,university.degree,0.0,0.0,0.0,telephone,614,2,999,0,NONEXISTENT,-2.9,"92,963","-40,8","1,268","5076,2",no,8-julio-2015,42.816,-91.954,7287cb94-c055-4904-9dbf-15469e4ea00c,131382,2,0,2014-01-31,31
22131,NaN,retired,MARRIED,basic.4y,0.0,0.0,0.0,cellular,123,1,999,0,NONEXISTENT,1.4,"93,444","-36,1","4,963","5228,1",no,28-junio-2019,31.011,-76.518,837e6bd3-c340-45f2-8143-97ff8cdcc6a2,130148,0,2,2013-05-15,4
24905,33.0,services,SINGLE,high.school,0.0,0.0,0.0,cellular,105,1,999,1,FAILURE,-0.1,"93,2",-42,NaN,"5195,8",no,6-octubre-2015,35.756,-98.852,73d037c3-5697-4685-9ff5-df5b389416a1,72722,1,0,2013-05-11,11


----------
DIMENSIONES
El conjunto de datos presenta 43000 filas y 28 columnas
----------
INFORMACION DE COLUMNAS
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 28 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   age                37880 non-null  float64       
 1   job                42655 non-null  object        
 2   marital            42915 non-null  object        
 3   education          41193 non-null  object        
 4   default            34019 non-null  float64       
 5   housing            41974 non-null  float64       
 6   loan               41974 non-null  float64       
 7   contact            43000 non-null  object        
 8   duration           43000 non-null  int64         
 9   campaign           43000 non-null  int64         
 10  pdays              43000 non-null  int64         
 11  previous           43000 non-null  int64         
 12  po

None

----------
DUPLICADOS
Tenemos 0 duplicados
----------
ELEMENTOS NULOS


age                  11.906977
job                   0.802326
marital               0.197674
education             4.202326
default              20.886047
housing               2.386047
loan                  2.386047
contact               0.000000
duration              0.000000
campaign              0.000000
pdays                 0.000000
previous              0.000000
poutcome              0.000000
emp.var.rate          0.000000
cons.price.idx        1.095349
cons.conf.idx         0.000000
euribor3m            21.525581
nr.employed           0.000000
y                     0.000000
date                  0.576744
latitude              0.000000
longitude             0.000000
id_                   0.000000
Income                0.000000
Kidhome               0.000000
Teenhome              0.000000
Dt_Customer           0.000000
NumWebVisitsMonth     0.000000
dtype: float64

----------
FRECUENCIA CATEGORÍAS
job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
Name: count, dtype: int64
-----------
marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
Name: count, dtype: int64
-----------
education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4y                4356
basic.6y                2386
illiterate                18
Name: count, dtype: int64
-----------
contact
cellular     27396
telephone    15604
Name: count, dtype: int64
-----------
poutcome
NONEXISTENT    37103
FAILURE         4461
SUCCESS         1436
Name: count, dtype: int64
-----------
cons.price.idx
93,994    7938
93,918    6937
92,893    5985
93,444    5349
94,465    4522
93,2      3731
93,075    2552
92,201 